### ISSUE TITLE
spack reindex invalid ref_count

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#1320]
```
$ git pull upstream develop
From github.com:LLNL/spack
 * branch            develop    -> FETCH_HEAD
Already up-to-date.
$ spack reindex
Traceback (most recent call last):
  File "/soft/spack/bin/spack", line 179, in <module>
    main()
  File "/soft/spack/bin/spack", line 157, in main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack/lib/spack/spack/cmd/reindex.py", line 31, in reindex
    spack.installed_db.reindex(spack.install_layout)
  File "/blues/gpfs/home/software/spack/lib/spack/spack/database.py", line 321, in reindex
    self._check_ref_counts()
  File "/blues/gpfs/home/software/spack/lib/spack/spack/database.py", line 353, in _check_ref_counts
    (key, found, expected, self._index_path))
AssertionError: Invalid ref_count: gts26usyvz7kzkoywzqsdpe3gs5qrson: 2 (expected 3), in DB /blues/gpfs/home/software/spack/opt/spack/.spack-db/index.yaml
```
Help...

**Comment by eschnett** [comment]
See also #1265...

I "solved" this by deleting my `opt` directory and re-installing everything from scratch.

**Comment by adamjstewart** [comment]
Trust me, I would love to do that. But I have close to 500 packages installed that my users depend on. By the way, most of those packages are duplicates since Spack reinstalls everything whenever a new version is added, a new variant is added, a new hash scheme is used, or whenever it just gets bored.

**Comment by alalazo** [comment]
@adamjstewart Poor man solution:
```
$ mv opt/spack/.spack-db/ opt/spack/.spack-db-bkp/ # Just in case...

$ spack reindex

$ spack find
==> 25 installed packages.
-- linux-Ubuntu14-x86_64 / gcc@4.8 ------------------------------
astyle@2.04    automake@1.15  cmake@3.5.2  hdf5@1.8.16  hwloc@1.11.3  libpciaccess@0.13.4  libtool@2.4.6  lua@5.3.2                lua-luaposix@33.4.0  ncurses@6.0   openmpi@1.10.3   readline@6.3  zlib@1.2.8
autoconf@2.69  blitz@1.0.0    gmp@6.1.0    hdf5@1.8.16  libffi@3.2.1  libsigsegv@2.10      lmod@6.4.1     lua-luafilesystem@1_6_3  m4@1.4.17            openjpeg@2.1  pkg-config@0.28  szip@2.1

```

**Comment by tgamblin** [comment]
@adamjstewart: does @alalazo's solution work for you?  I do not know how else to debug this without a reproducer, although we could make the database more resilient to bad ref counts and have it try to do something more sensible.

**Comment by adamjstewart** [comment]
I didn't get a chance to try @alalazo's solution. I figured my problem had to do with my dozens of identical python installations, so I uninstalled a bunch of them and `spack reindex` works again. A possible cause may be manually uninstalling a Python package without deactivating it first. I've had to do this a lot lately, for the same reason as discussed in #1178. I'm now in the process of reinstalling and consolidating all of my python installations. Eventually I would like to reinstall everything I had before the newarch merge destroyed all of my hashes.

**Comment by TheQueasle** [comment]
I've hit a similar issue:
```
[user@build-server]$ spack reindex
Traceback (most recent call last):
  File "/cluster/rcss-spack/bin/spack", line 184, in <module>
    main()
  File "/cluster/rcss-spack/bin/spack", line 161, in main
    return_val = command(parser, args)
  File "/cluster/rcss-spack/lib/spack/spack/cmd/reindex.py", line 31, in reindex
    spack.installed_db.reindex(spack.install_layout)
  File "/cluster/rcss-spack/lib/spack/spack/database.py", line 322, in reindex
    self._check_ref_counts()
  File "/cluster/rcss-spack/lib/spack/spack/database.py", line 354, in _check_ref_counts
    (key, found, expected, self._index_path))
AssertionError: Invalid ref_count: 7px7veadsi5qx47g4tr5pflxtxokaa7s: 1 (expected 2), in DB /cluster/rcss-spack/opt/spack/.spack-db/index.yaml
``` 

Attempted the fix @alalazo suggested, but this did not work.

I then moved all python packages (including python itself) out of `opt/` and started moving them back in one at a time then running the reindex.  After moving `py-matplotlib` back into `opt/`, the reindex fails.

**Comment by adamjstewart** [comment]
I would try removing all of your Python installations and installing them again with Spack. I've had a lot of problems when trying to update Python packages. I assume you are also using `spack activate`?

**Comment by TheQueasle** [comment]
Actually, I've never used `spack activate`.

This is a fresh install of spack with code pulled on 08/18/16.

I'll try your suggestion of uninstalling all python packages and reinstalling them.

**Comment by eschnett** [comment]
The `spack reindex` problems appear on fresh installs, where there were no problems at all during the install. They are there right away after the install.

**Comment by adamjstewart** [comment]
Closed by #1688
